app/assets/src/components/views/discovery/mapping/MapPlayground.jsx

```diff
@@ -1,78 +1,153 @@
+      if (match) {
+        let [lat, lon] = this.parseLatLon(match[1], match[2]);
+        if (!(lat && lon)) return;
+        locationName = `${lat}, ${lon}`;
+        const item = {
           name: result.name,
           id: result.id
         };
-        if (locationsToSamples.hasOwnProperty(loc)) {
-          locationsToSamples[loc].push(formatted);
+
+        // Group items under the location name
+        if (locationsToItems.hasOwnProperty(locationName)) {
+          locationsToItems[locationName].items.push(item);
         } else {
-          locationsToSamples[loc] = [formatted];
+          locationsToItems[locationName] = {
+            lat: lat,
+            lon: lon,
+            items: [item]
+          };
         }
       }
     });
 
     this.state = {
-      locationsToSamples,
-      viewport: {}
+      locationsToItems: locationsToItems,
+      viewport: {},
+      popups: []
     };
   }
 
+  parseLatLon = (lat, lon) => {
+    // Round the coordinates for some minimal aggregation
+    lat = parseFloat(parseFloat(lat).toFixed(2));
+    lon = parseFloat(parseFloat(lon).toFixed(2));
+    // Reject invalid coordinates
+    if (lat < -90 || lat > 90 || lon < -180 || lon > 180) {
+      return [null, null];
+    } else {
+      return [lat, lon];
+    }
+  };
+
   updateViewport = viewport => {
     this.setState({ viewport });
   };
 
-  renderMarker = (markerData, index) => {
+  renderMarker = (marker, index) => {
     const { viewport } = this.state;
-    const [lat, lon] = markerData[0].split(",");
-    const pointCount = markerData[1].length;
-    const minSize = 14;
+    const [locationName, markerData] = marker;
+    const pointCount = markerData.items.length;
+    const minSize = 12;
     // Scale based on the zoom and point count (zoomed-in = higher zoom)
     const markerSize = Math.max(
-      pointCount * (get("zoom", viewport) || 3),
+      Math.log(pointCount) / Math.log(1.5) * (get("zoom", viewport) || 3),
       minSize
     );
+    const hoverContent = (
+      <div>
+        <div className={cs.title}>{locationName}</div>
+        <div className={cs.description}>{`Samples: ${pointCount}`}</div>
+      </div>
+    );
+    const popupInfo = {
+      name: locationName,
+      lat: markerData.lat,
+      lon: markerData.lon,
+      markerIndex: index,
+      items: markerData.items
+    };
     return (
       <Marker
         key={`marker-${index}`}
-        latitude={parseFloat(lat)}
-        longitude={parseFloat(lon)}
+        latitude={markerData.lat}
+        longitude={markerData.lon}
       >
         <CircleMarker
           size={markerSize}
-          hoverContent={`${pointCount} sample${pointCount > 1 ? "s" : ""}`}
+          hoverContent={hoverContent}
+          onClick={() => this.openPopup(popupInfo)}
         />
       </Marker>
     );
   };
 
+  openPopup(popupInfo) {
+    const popups = Object.assign([], this.state.popups);
+    popups.push(popupInfo);
+    this.setState({ popups });
```

Comment:
You can use `concat` from lodash/fp here and simplify to one expression, same as below.

Reply:
👍 changed